### ISSUE TITLE
SmallRyeConfigBuilder does not load addDiscoveredConverters after 3.8.0.CR1

### ISSUE THREAD

**Issue by ikovalyov** [quarkusio/quarkus#38887]
### Describe the bug

Previouly converters were always discovered during runtime execution phase and now they are missed

### Expected behavior

Converters are still doscovered in runtime

### Actual behavior

Converters are not discovered.

```
java.lang.IllegalArgumentException: SRCFG02007: No Converter registered for interface ***
		at io.smallrye.config.inject.ConfigProducerUtil.lambda$resolveConverter$3(ConfigProducerUtil.java:183)
		at java.base/java.util.Optional.orElseThrow(Optional.java:403)
		at io.smallrye.config.inject.ConfigProducerUtil.resolveConverter(ConfigProducerUtil.java:183)
		at io.smallrye.config.inject.ConfigProducerUtil.getValue(ConfigProducerUtil.java:100)
		at io.quarkus.arc.runtime.ConfigRecorder.validateConfigProperties(ConfigRecorder.java:60)
		... 10 more
```

### How to Reproduce?

1. Create converter classes in a build step using quarkus gizmo.
2. Put a reference to this class in a META-INF\services\org.eclipse.microprofile.config.spi.Converter file.
3. Try to inject object for which converter was registered. Generated files are discovered by SmallRyeConfig in build time but not discovered in runtime.

**Comment by gsmet** [comment]
You're mentioning 3.8.0.CR1, you have the same issue with 3.7.3, right?

It would help if you could provide a simple reproducer. Thanks!

/cc @radcortez this looks concerning

**Comment by ikovalyov** [comment]
Based on my discoveres in version 3.2.6 `ConfigUtils` was creating `SmallRyeConfigBuilder` with `addDiscoveredConverters` being called in both build and run time.
In new version There are separate builders for build and run time and run time builder creates `SmallRyeCOnfigBuilder` without call to `addDiscoveredConverters`

**Comment by gsmet** [comment]
@radcortez FWIW, the important point here is that the code of the converter is generated with Gizmo.

@ikovalyov I think you should at least post how the converter is generated.

**Comment by radcortez** [comment]
Yes, we have the `addDiscovered*` to false in runtime to avoid the `ServiceLoader`. We discover the services during build time and register them directly (generated) with the builder.

That piece is done here:
https://github.com/quarkusio/quarkus/blob/6053da3880cdbb5a53092b2c320be89cad77f7a8/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java#L194-L282

Maybe your generation is happening after this config build step and is not being picked up?

**Comment by ikovalyov** [comment]
Hi!
Yes, our code is generated on a runtime.

**Comment by radcortez** [comment]
> We indeed see it is triggered after executing code you mentioned.

Ok, so in that case, it is not going to be discovered by our code. 

My recommendation is to directly register the converter with a `RunTimeConfigBuilderBuildItem` and a `io.quarkus.runtime.configuration.ConfigBuilder`. This may require some additional code generation because the `withConverter` requires a new instance of the registered `Converter`.

Or just turn the discovery on again in the builder, with the downside of extra operations in the native image. You probably need to add the service loader file as the native image resource manually. We don't do it automatically, because we determine everything at build time and register the converters directly to avoid that.

**Comment by ikovalyov** [comment]
We were able to resolve this issue using RunTimeConfigBuilderBuildItem where we enabled back the `addDiscoveredConverters` option.

Important notice: while it appeared our generators were executed before `generateBuilders` action quarkus could not load converters created by our extension. Based on other tickets here I was afraid we would need to move code generation out of the quarkus and then inject them as dependency but happily it works as it is.

Thank you.
